File: frontend/src/api/questionBank.ts
import api from './auth'

// ===== Subject / Domain / Point =====

export interface SubjectItem {
  id: string
  name: string
  description: string | null
  sort_order: number
  created_at: string
  updated_at: string
  domains: DomainItem[]
}

export interface DomainItem {
  id: string
  subject_id: string
  name: string
  description: string | null
  sort_order: number
  created_at: string
  updated_at: string
  knowledge_points: KnowledgePointItem[]
}

export interface KnowledgePointItem {
  id: string
  domain_id: string
  name: string
  description: string | null
  difficulty: number
  sort_order: number
  created_at: string
  updated_at: string
}

// ===== Question Bank (PRD v3) =====

export interface BankItem {
  id: string
  owner_id: string
  subject_id: string
  name: string
  description: string | null
  visibility: string
  total_questions: number
  tags: string[]
  created_at: string
  updated_at: string
}

// ===== Question (PRD v3) =====
// 存储说明:
// - content: { stem: "题干", options?: [{key: "A", text: "选项"}], ... }
// - answer: { correct_answer: ["B"], explanation: "解析", ... }
// - knowledge_point_uuids: Neo4j 知识点 UUID 列表

export type QuestionType =
  | 'single_choice' | 'multiple_choice' | 'fill_blank'
  | 'true_false' | 'short_answer' | 'programming' | 'essay'

export type QuestionDifficulty =
  | 'beginner' | 'basic' | 'intermediate' | 'advanced' | 'competition'

export interface QuestionItem {
  id: string
  bank_id: string
  type: QuestionType
  content: { stem: string; options?: { key: string; text: string }[]; code_template?: string; images?: string[] }
  answer: { correct_answer: string[]; explanation: string; difficulty_rationale?: string; suggested_time_seconds?: number }
  difficulty: QuestionDifficulty
  priority: number
  knowledge_point_uuids: string[]
  tags: string[]
  ai_generated: boolean
  source: string
  status: string
  created_by: string | null
  created_at: string
  updated_at: string
}

export const questionBankApi = {
  // === Subjects ===
  listSubjects: () =>
    api.get<{ subjects: SubjectItem[]; total: number }>('/question-bank/subjects'),
  getSubject: (id: string) =>
    api.get<SubjectItem>(`/question-bank/subjects/${id}`),
  createSubject: (data: { name: string; description?: string; sort_order?: number }) =>
    api.post<SubjectItem>('/question-bank/subjects', data),
  updateSubject: (id: string, data: { name: string; description?: string; sort_order?: number }) =>
    api.put<SubjectItem>(`/question-bank/subjects/${id}`, data),
  deleteSubject: (id: string) =>
    api.delete(`/question-bank/subjects/${id}`),

  // === Domains ===
  listDomains: (subjectId: string) =>
    api.get<DomainItem[]>(`/question-bank/subjects/${subjectId}/domains`),
  createDomain: (subjectId: string, data: { name: string; description?: string; sort_order?: number }) =>
    api.post<DomainItem>(`/question-bank/subjects/${subjectId}/domains`, data),
  updateDomain: (id: string, data: { name: string; description?: string; sort_order?: number }) =>
    api.put<DomainItem>(`/question-bank/domains/${id}`, data),
  deleteDomain: (id: string) =>
    api.delete(`/question-bank/domains/${id}`),

  // === Knowledge Points ===
  listPoints: (domainId: string) =>
    api.get<KnowledgePointItem[]>(`/question-bank/domains/${domainId}/points`),
  createPoint: (domainId: string, data: { name: string; description?: string; difficulty?: number; sort_order?: number }) =>
    api.post<KnowledgePointItem>(`/question-bank/domains/${domainId}/points`, data),
  updatePoint: (id: string, data: { name: string; description?: string; difficulty?: number; sort_order?: number }) =>
    api.put<KnowledgePointItem>(`/question-bank/points/${id}`, data),
  deletePoint: (id: string) =>
    api.delete(`/question-bank/points/${id}`),

  // === Banks ===
  listBanks: (params?: { subject_id?: string; search?: string; page?: number; page_size?: number }) =>
    api.get<{ banks: BankItem[]; total: number }>('/question-bank/banks', { params }),
  createBank: (data: { name: string; subject_id: string; description?: string; visibility?: string; tags?: string[] }) =>
    api.post<BankItem>('/question-bank/banks', data),
  getBank: (id: string) =>
    api.get<BankItem>(`/question-bank/banks/${id}`),
  updateBank: (id: string, data: { name?: string; description?: string; visibility?: string; tags?: string[] }) =>
    api.put<BankItem>(`/question-bank/banks/${id}`, data),
  deleteBank: (id: string) =>
    api.delete(`/question-bank/banks/${id}`),

  // === Questions ===
  listQuestions: (bankId: string, params?: {
    type?: string; difficulty?: string; status?: string;
    knowledge_point_uuid?: string; search?: string; page?: number; page_size?: number
  }) =>
    api.get<{ questions: QuestionItem[]; total: number; page: number; page_size: number }>(
      `/question-bank/banks/${bankId}/questions`, { params }),
  createQuestion: (bankId: string, data: {
    type: string; content: Record<string, any>; answer?: Record<string, any>;
    difficulty?: string; priority?: number; knowledge_point_uuids?: string[];
    tags?: string[]; status?: string; ai_generated?: boolean; source?: string
  }) =>
    api.post<QuestionItem>(`/question-bank/banks/${bankId}/questions`, data),
  getQuestion: (id: string) =>
    api.get<QuestionItem>(`/question-bank/questions/${id}`),
  updateQuestion: (id: string, data: Record<string, any>) =>
    api.put<QuestionItem>(`/question-bank/questions/${id}`, data),
  deleteQuestion: (id: string) =>
    api.delete(`/question-bank/questions/${id}`),
  regenerateQuestion: (id: string, feedback: string) =>
    api.post<QuestionItem>(`/question-bank/questions/${id}/regenerate`, { feedback }),

  // === AI Generate ===
  aiGenerate: (bankId: string, data: {
    message: string; conversation_history?: { role: string; content: string }[];
    collected_params?: Record<string, any>;
  }) => api.post<{
    reply: string; collected_params: Record<string, any>;
    generated_questions: any[]; is_complete: boolean;
  }>(`/question-bank/banks/${bankId}/ai-generate`, data),

  // === AI Generate Stream (SSE) ===
  aiGenerateStream: (
    bankId: string,
    data: {
      message: string; conversation_history?: { role: string; content: string }[];
      collected_params?: Record<string, any>;
    },
    callbacks: {
      onChunk: (text: string) => void;
      onComplete: (result: { reply: string; generated_questions: any[]; is_complete: boolean; collected_params?: Record<string, any> }) => void;
      onError: (error: string) => void;
    },
  ) => {
    const token = localStorage.getItem('access_token')
    return fetch(`/api/v1/question-bank/banks/${bankId}/ai-generate-stream`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    }).then(async (response) => {
      if (!response.ok) {
        callbacks.onError(`请求失败: ${response.status}`)
        return
      }
      const reader = response.body!.getReader()
      const decoder = new TextDecoder()
      let buffer = ''

      while (true) {
        const { done, value } = await reader.read()
        if (done) break
        buffer += decoder.decode(value, { stream: true })

        const lines = buffer.split('\n')
        buffer = lines.pop() || ''

        for (const line of lines) {
          const trimmed = line.trim()
          if (!trimmed.startsWith('data: ')) continue
          try {
            const event = JSON.parse(trimmed.slice(6))
            switch (event.type) {
              case 'chunk':
                callbacks.onChunk(event.content)
                break
              case 'complete':
                callbacks.onComplete(event)
                break
              case 'error':
                callbacks.onError(event.content)
                break
            }
          } catch { /* skip malformed events */ }
        }
      }
    }).catch((err) => {
      callbacks.onError(err.message || '网络错误')
    })
  },

  // === AI Context Persistence ===
  getAIContext: (bankId: string) =>
    api.get<{ has_context: boolean; history: { role: string; content: string }[]; collected_params: Record<string, any>; generated_questions: any[] }>(
      `/question-bank/banks/${bankId}/ai-context`),
  clearAIContext: (bankId: string) =>
    api.delete(`/question-bank/banks/${bankId}/ai-context`),
  updateSavedQuestions: (bankId: string, savedIndices: number[]) =>
    api.post(`/question-bank/banks/${bankId}/ai-context/saved-questions`, { saved_indices: savedIndices }),

  // === Domain Counts ===
  getDomainCounts: (bankId: string, config: {
    question_types?: string[]
    domain_ids?: string[]
    knowledge_point_uuids?: string[]
    only_unanswered?: boolean
    only_wrong?: boolean
    only_error_prone?: boolean
    answer_mode?: string
  }) => api.post<{ domain_id: string; domain_name: string; total: number; unanswered: number; wrong: number }[]>(
    `/question-bank/banks/${bankId}/domain-counts`, config),

  // === Practice Sessions ===
  createPracticeSession: (bankId: string, data: {
    mode?: string; answer_mode?: string; question_order?: string[]
  }) => api.post<{
    id: string; bank_id: string; mode: string; status: string;
    question_order: string[]; stats: Record<string, any>; answer_mode: string;
    started_at: string; finished_at: string | null;
  }>(`/question-bank/banks/${bankId}/practice-sessions`, data),

  updatePracticeSession: (sessionId: string, data: {
    status?: string; current_index?: number; stats?: Record<string, any>; finished_at?: string
  }) => api.put<Record<string, any>>(`/question-bank/practice-sessions/${sessionId}`, data),

  listPracticeSessions: (params?: { bank_id?: string; page?: number; page_size?: number }) =>
    api.get<{ sessions: any[]; total: number }>('/question-bank/practice-sessions', { params }),

  getPracticeSession: (sessionId: string) =>
    api.get<Record<string, any>>(`/question-bank/practice-sessions/${sessionId}`),

  getPracticeSessionQuestions: (sessionId: string) =>
    api.get<QuestionItem[]>(`/question-bank/practice-sessions/${sessionId}/questions`),

  // === Knowledge Point → Practice Bank ===
  getKnowledgePointPracticeBank: (kpUuid: string) =>
    api.get<{ bank_id: string; bank_name: string }>(
      `/question-bank/knowledge-points/${kpUuid}/practice-bank`),

  // === Practice ===
  startPractice: (bankId: string, config: {
    time_limit_minutes?: number | null
    question_count?: number | null
    question_types?: string[]
    domain_ids?: string[]
    knowledge_point_uuids?: string[]
    only_unanswered?: boolean
    only_wrong?: boolean
    only_error_prone?: boolean
    answer_mode?: string
  }) => api.post<QuestionItem[]>(
    `/question-bank/banks/${bankId}/practice-questions`, config),

  submitAnswer: (questionId: string, data: {
    answer_content: Record<string, any>
    is_correct: boolean
    time_spent_seconds?: number | null
    session_id?: string | null
  }) => api.post<{
    is_correct: boolean
    recommended_resources?: Array<{ id: string; title: string; resource_type: string; knowledge_points: string[] }>
  }>(`/question-bank/questions/${questionId}/submit-answer`, data),

  submitAnswers: (bankId: string, data: {
    answers: { question_id: string; answer_content: Record<string, any>; is_correct: boolean; time_spent_seconds?: number | null; session_id?: string | null }[]
  }) => api.post<{
    results: Array<{
      is_correct: boolean
      recommended_resources?: Array<{ id: string; title: string; resource_type: string; knowledge_points: string[] }>
    }>
  }>(`/question-bank/banks/${bankId}/submit-answers`, data),

  // === Exam Papers ===
  listExamPapers: (bankId: string, params?: { page?: number; page_size?: number }) =>
    api.get<{ papers: any[]; total: number }>(`/question-bank/banks/${bankId}/exam-papers`, { params }),

  getExamPaper: (paperId: string) =>
    api.get<any>(`/question-bank/exam-papers/${paperId}`),

  createExamPaper: (bankId: string, data: {
    title: string; description?: string; total_score?: number; time_limit_minutes?: number | null;
    generate_method?: string; sections: any[];
  }) => api.post<any>(`/question-bank/banks/${bankId}/exam-papers`, data),

  deleteExamPaper: (paperId: string) =>
    api.delete(`/question-bank/exam-papers/${paperId}`),

  suggestQuestions: (bankId: string, data: {
    sections: { name: string; question_type: string; count: number; score_per_question: number; difficulty?: string | null; domain_ids: string[] }[];
    exclude_question_ids?: string[];
    deterministic?: boolean;
    seed_only?: boolean;
  }) => api.post<{ sections: any[]; total_questions: number }>(
    `/question-bank/banks/${bankId}/exam-papers/suggest-questions`, data),

  parseUpload: (bankId: string, file: File) => {
    const fd = new FormData()
    fd.append('file', file)
    return api.post<{ filename: string; suggested_title: string; parsed_sections: any[]; full_text: string }>(
      `/question-bank/banks/${bankId}/exam-papers/parse-upload`, fd,
      { headers: { 'Content-Type': 'multipart/form-data' } })
  },

  aiGenerateExamQuestions: (bankId: string, data: {
    message: string; conversation_history?: { role: string; content: string }[];
    collected_params?: Record<string, any>;
  }) => api.post<{
    reply: string; collected_params: Record<string, any>;
    generated_questions: any[]; is_complete: boolean;
  }>(`/question-bank/banks/${bankId}/exam-papers/ai-generate`, data),

  exportExamPDF: (paperId: string) =>
    api.get(`/question-bank/exam-papers/${paperId}/export/pdf`, { responseType: 'blob' }),

  exportExamWord: (paperId: string) =>
    api.get(`/question-bank/exam-papers/${paperId}/export/word`, { responseType: 'blob' }),

  startExamPractice: (paperId: string) =>
    api.post<{ session_id: string; practice_url: string }>(
      `/question-bank/exam-papers/${paperId}/start-practice`),

  // === Error-Prone Tags (F5) ===
  listErrorProneQuestions: (params?: { bank_id?: string; page?: number; page_size?: number }) =>
    api.get<{ questions: QuestionItem[]; total: number }>('/question-bank/questions/error-prone', { params }),

  tagErrorProne: (questionId: string) =>
    api.post<{ question_id: string; tags: string[] }>(`/question-bank/questions/${questionId}/tag-error-prone`),

  untagErrorProne: (questionId: string) =>
    api.delete<{ question_id: string; tags: string[] }>(`/question-bank/questions/${questionId}/untag-error-prone`),

  // === Wrong Answer Book ===
  listWrongAnswers: (params?: {
    bank_id?: string; subject_id?: string; domain_id?: string;
    page?: number; page_size?: number
  }) => api.get<{ items: WrongAnswerItem[]; total: number }>(
    '/question-bank/wrong-answers', { params }),

  removeWrongAnswer: (recordId: string) =>
    api.delete(`/question-bank/wrong-answers/${recordId}`),

  getWrongReview: (bankId: string) =>
    api.get<WrongReviewResponse>(`/question-bank/banks/${bankId}/wrong-review`),

  diagnoseWrongAnswer: (recordId: string) =>
    api.post<{
      error_type: string; error_type_label: string;
      root_cause: string; suggestions: string[]; recommended_action: string
    }>(`/question-bank/wrong-answers/${recordId}/diagnose`),

  generatePracticeFromWrongAnswers: (params?: {
    bank_id?: string; question_count?: number; domain_ids?: string[]
  }) => api.post<QuestionItem[]>(
    '/question-bank/wrong-answers/generate-practice', null, { params }),

  // === Session Answers (Test History Detail) ===
  getSessionAnswers: (sessionId: string) =>
    api.get<{ session_id: string; items: SessionAnswerItem[]; total: number }>(
      `/question-bank/practice-sessions/${sessionId}/answers`),

  // === Self-Grade ===
  submitSelfGrade: (answerId: string, selfGrade: number) =>
    api.post<{ answer_id: string; self_grade: number }>(
      `/question-bank/answers/${answerId}/self-grade`, { self_grade: selfGrade }),

  submitBatchSelfGrade: (sessionId: string, grades: { answer_id: string; self_grade: number }[]) =>
    api.post<{ session_id: string; updated: { answer_id: string; self_grade: number }[] }>(
      `/question-bank/sessions/${sessionId}/self-grade`, { grades }),

  // === Daily Stats ===
  getDailyStats: (params?: {
    bank_id?: string; mode?: string; days?: number;
    start_date?: string; end_date?: string
  }) => api.get<{ items: DailyStatsItem[]; total: number }>(
    '/question-bank/daily-stats', { params }),

  // === Domain Completion ===
  getDomainCompletion: (domainId: string) =>
    api.get<DomainCompletion>(`/question-bank/domains/${domainId}/completion`),

  markDomainMastered: (domainId: string) =>
    api.post<MarkMasteredResponse>(`/question-bank/domains/${domainId}/mark-mastered`),
}

// === Domain Completion Types ===
export interface DomainCompletion {
  domain_id: string
  domain_name: string
  total_questions: number
  answered_questions: number
  wrong_count: number
  all_done: boolean
}

export interface MarkMasteredResponse {
  success: boolean
  marked_count: number
  message: string
}

// ===== Wrong Answer Book Types =====

export interface WrongAnswerItem {
  id: string
  question_id: string
  bank_id: string
  wrong_count: number
  first_wrong_at: string
  last_wrong_at: string
  question: QuestionItem
}

// === Wrong Review ===

export interface WrongReviewItem {
  question_id: string
  stem: string
  type: string
  options: Record<string, string> | null
  user_answer: string
  correct_answer: string[]
  explanation: string | null
  knowledge_points: string[]
  wrong_count: number
  last_wrong_at: string
}

export interface WrongReviewResponse {
  wrong_records: WrongReviewItem[]
  total: number
  bank_name: string
}

export interface SessionAnswerItem {
  answer_id: string
  question_id: string
  question: QuestionItem | null
  answer_content: Record<string, any>
  is_correct: boolean
  self_grade: number | null
  time_spent_seconds: number | null
  answered_at: string | null
}

export interface DailyStatsItem {
  date: string
  total_questions: number
  correct_count: number
  incorrect_count: number
  total_time_spent_seconds: number
  session_count: number
  accuracy: number
}

File: app/schemas/question_bank.py
from pydantic import BaseModel, Field, field_validator
from typing import Optional, List, Dict, Any
from uuid import UUID
from datetime import datetime


# ===== Subject / KnowledgeDomain / KnowledgePoint =====

class KnowledgePointCreate(BaseModel):
    name: str
    description: Optional[str] = None
    difficulty: int = 1
    sort_order: int = 0


class KnowledgePointResponse(BaseModel):
    id: UUID
    domain_id: UUID
    name: str
    description: Optional[str] = None
    difficulty: int
    sort_order: int
    created_at: datetime
    updated_at: datetime
    model_config = {"from_attributes": True}


class KnowledgeDomainCreate(BaseModel):
    name: str
    description: Optional[str] = None
    sort_order: int = 0


class KnowledgeDomainResponse(BaseModel):
    id: UUID
    subject_id: UUID
    name: str
    description: Optional[str] = None
    sort_order: int
    created_at: datetime
    updated_at: datetime
    knowledge_points: List[KnowledgePointResponse] = []
    model_config = {"from_attributes": True}


class SubjectCreate(BaseModel):
    name: str
    description: Optional[str] = None
    sort_order: int = 0


class SubjectResponse(BaseModel):
    id: UUID
    name: str
    description: Optional[str] = None
    cover_image: Optional[str] = None
    creator_id: Optional[UUID] = None
    sort_order: int
    created_at: datetime
    updated_at: datetime
    domains: List[KnowledgeDomainResponse] = []
    model_config = {"from_attributes": True}


class SubjectListResponse(BaseModel):
    subjects: List[SubjectResponse]
    total: int


# ===== Question Bank =====

class QuestionBankCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    subject_id: UUID
    description: Optional[str] = None
    visibility: str = "private"
    tags: List[str] = []


class QuestionBankUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    visibility: Optional[str] = None
    tags: Optional[List[str]] = None


class QuestionBankResponse(BaseModel):
    id: UUID
    owner_id: UUID
    subject_id: UUID
    name: str
    description: Optional[str] = None
    visibility: str = "private"
    total_questions: int = 0
    tags: List[str] = []
    created_at: datetime
    updated_at: datetime
    model_config = {"from_attributes": True}


class QuestionBankListResponse(BaseModel):
    banks: List[QuestionBankResponse]
    total: int


# ===== Question =====

class AnswerSchema(BaseModel):
    correct_answer: List[str] = Field(default_factory=list)
    explanation: str = ""
    difficulty_rationale: str = ""
    suggested_time_seconds: int = 60


class QuestionCreate(BaseModel):
    type: str = Field(..., pattern=r"^(single_choice|multiple_choice|fill_blank|true_false|short_answer|programming|essay)$")
    content: Dict[str, Any] = Field(default_factory=dict)
    answer: Dict[str, Any] = Field(default_factory=dict)
    difficulty: str = "basic"
    priority: int = 0
    knowledge_point_uuids: List[str] = Field(default_factory=list)
    tags: List[str] = []
    ai_generated: bool = False
    source: str = "manual"
    status: str = "published"


class QuestionUpdate(BaseModel):
    type: Optional[str] = None
    content: Optional[Dict[str, Any]] = None
    answer: Optional[Dict[str, Any]] = None
    difficulty: Optional[str] = None
    priority: Optional[int] = None
    knowledge_point_uuids: Optional[List[str]] = None
    tags: Optional[List[str]] = None
    status: Optional[str] = None


class QuestionResponse(BaseModel):
    id: UUID
    bank_id: UUID
    type: str
    content: Dict[str, Any]
    answer: Dict[str, Any]
    difficulty: str
    priority: int
    knowledge_point_uuids: List[str] = []
    tags: List[str] = []
    ai_generated: bool = False
    source: str = "manual"
    status: str = "published"
    created_by: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime
    model_config = {"from_attributes": True}

    @field_validator("tags", mode="before")
    @classmethod
    def coerce_tags(cls, v):
        return v if v is not None else []

    @field_validator("ai_generated", mode="before")
    @classmethod
    def coerce_ai_generated(cls, v):
        return False if v is None else v

    @field_validator("source", mode="before")
    @classmethod
    def coerce_source(cls, v):
        return "manual" if v is None else v


class QuestionListResponse(BaseModel):
    questions: List[QuestionResponse]
    total: int
    page: int
    page_size: int


# ===== Practice =====

class DomainQuestionCount(BaseModel):
    """章节题目计数"""
    domain_id: str
    domain_name: str
    total: int
    unanswered: int = 0
    wrong: int = 0


class PracticeConfig(BaseModel):
    """自测配置"""
    time_limit_minutes: Optional[int] = Field(None, description="时间限制（分钟），null=无限制")
    question_count: Optional[int] = Field(None, ge=1, le=200, description="题目数量，null=全部")
    question_types: List[str] = Field(default_factory=list, description="题型过滤，空=全部")
    domain_ids: List[str] = Field(default_factory=list, description="章节ID过滤，空=全部")
    knowledge_point_uuids: List[str] = Field(default_factory=list, description="知识点UUID过滤，空=全部")
    only_unanswered: bool = Field(False, description="只做未做过的")
    only_wrong: bool = Field(False, description="只做错题")
    only_error_prone: bool = Field(False, description="仅易错题")
    answer_mode: str = Field("during", description="答案显示模式: during=边测边看, after=测后看答案")


class PracticeQuestionResponse(BaseModel):
    """练习用题目"""
    id: UUID
    type: str
    content: Dict[str, Any]
    answer: Dict[str, Any] = Field(default_factory=dict)
    difficulty: str
    knowledge_point_uuids: List[str] = []
    tags: List[str] = []
    model_config = {"from_attributes": True}


class AnswerSubmitRequest(BaseModel):
    """单题答案提交"""
    answer_content: Dict[str, Any] = Field(default_factory=dict)
    is_correct: bool = False
    time_spent_seconds: Optional[int] = None
    session_id: Optional[str] = Field(None, description="练习会话ID")


class AnswerSubmitResponse(BaseModel):
    is_correct: bool
    recommended_resources: List[dict] = Field(default_factory=list, description="答错时推荐的已有学习资源")


class BatchAnswerItem(BaseModel):
    question_id: str
    answer_content: Dict[str, Any] = Field(default_factory=dict)
    is_correct: bool = False
    time_spent_seconds: Optional[int] = None
    session_id: Optional[str] = None


class BatchAnswerSubmitRequest(BaseModel):
    answers: List[BatchAnswerItem]


class BatchAnswerSubmitResponse(BaseModel):
    results: List[AnswerSubmitResponse]


# ===== Self-Grade =====

class SelfGradeRequest(BaseModel):
    """单题自评提交"""
    self_grade: float = Field(..., ge=0.0, le=1.0, description="自评分数 0.0-1.0")


class SelfGradeItem(BaseModel):
    """批量自评中的单项"""
    answer_id: str
    self_grade: float = Field(..., ge=0.0, le=1.0)


class BatchSelfGradeRequest(BaseModel):
    """批量自评提交"""
    grades: List[SelfGradeItem]


# ===== Practice Session =====

class PracticeSessionCreate(BaseModel):
    """创建练习会话"""
    mode: str = "random"
    answer_mode: str = "during"
    question_order: List[str] = Field(default_factory=list)


class PracticeSessionUpdate(BaseModel):
    """更新练习会话"""
    status: Optional[str] = None
    current_index: Optional[int] = None
    stats: Optional[Dict[str, Any]] = None
    finished_at: Optional[datetime] = None


class PracticeSessionResponse(BaseModel):
    id: UUID
    user_id: UUID
    bank_id: UUID
    mode: str
    status: str
    question_order: List[str] = []
    current_index: int = 0
    stats: Dict[str, Any] = {}
    answer_mode: str = "during"
    started_at: Optional[datetime] = None
    finished_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    model_config = {"from_attributes": True}


class PracticeSessionListResponse(BaseModel):
    sessions: List[PracticeSessionResponse]
    total: int


# ===== Exam Paper =====

class ExamPaperSection(BaseModel):
    name: str
    question_type: str
    count: int = Field(ge=1, le=200)
    score_per_question: int = Field(1, ge=1, le=100)
    question_ids: List[str] = Field(default_factory=list)
    difficulty: Optional[str] = None
    domain_ids: List[str] = Field(default_factory=list)


class ExamPaperCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None
    total_score: int = Field(100, ge=1, le=1000)
    time_limit_minutes: Optional[int] = Field(None, ge=1, le=600)
    generate_method: str = "manual"
    sections: List[ExamPaperSection] = Field(default_factory=list)


class ExamPaperUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    total_score: Optional[int] = None
    time_limit_minutes: Optional[int] = None
    status: Optional[str] = None


class ExamPaperListItem(BaseModel):
    id: UUID
    bank_id: UUID
    title: str
    description: Optional[str] = None
    generate_method: str
    status: str
    total_questions: int
    total_score: int
    time_limit_minutes: Optional[int] = None
    created_at: datetime
    updated_at: datetime
    model_config = {"from_attributes": True}


class ExamPaperDetailResponse(BaseModel):
    id: UUID
    bank_id: UUID
    title: str
    description: Optional[str] = None
    config: Dict[str, Any] = {}
    generate_method: str
    status: str
    total_questions: int
    total_score: int
    time_limit_minutes: Optional[int] = None
    created_at: datetime
    updated_at: datetime
    sections: List[Dict[str, Any]] = []
    model_config = {"from_attributes": True}


class ExamPaperListResponse(BaseModel):
    papers: List[ExamPaperListItem]
    total: int


class SuggestQuestionsRequest(BaseModel):
    sections: List[ExamPaperSection]
    exclude_question_ids: List[str] = Field(default_factory=list)


class SuggestQuestionsResponse(BaseModel):
    sections: List[Dict[str, Any]]
    total_questions: int


class ParseUploadResponse(BaseModel):
    filename: str
    suggested_title: str
    parsed_sections: List[Dict[str, Any]] = Field(default_factory=list)
    full_text: str = ""


class AIGenerateExamResponse(BaseModel):
    generated_questions: List[Dict[str, Any]]
    total: int


class StartExamPracticeResponse(BaseModel):
    session_id: str
    practice_url: str


# ===== Wrong Answer Book =====

class WrongAnswerItem(BaseModel):
    id: UUID
    question_id: str
    bank_id: str
    wrong_count: int
    first_wrong_at: datetime
    last_wrong_at: datetime
    question: QuestionResponse
    model_config = {"from_attributes": True}


class WrongAnswerListResponse(BaseModel):
    items: List[WrongAnswerItem]
    total: int


# ===== Session Answers (Test History Detail) =====

class SessionAnswerItem(BaseModel):
    answer_id: str
    question_id: str
    question: Optional[QuestionResponse] = None
    answer_content: Dict[str, Any] = {}
    is_correct: bool
    self_grade: Optional[float] = None  # 自评分数 0.0-1.0
    time_spent_seconds: Optional[int] = None
    answered_at: Optional[datetime] = None


class SessionAnswerListResponse(BaseModel):
    session_id: str
    items: List[SessionAnswerItem]
    total: int


# ===== Daily Stats =====

class DailyStatsItem(BaseModel):
    date: str
    total_questions: int = 0
    correct_count: int = 0
    incorrect_count: int = 0
    total_time_spent_seconds: int = 0
    session_count: int = 0
    accuracy: float = 0.0


class DailyStatsResponse(BaseModel):
    items: List[DailyStatsItem]
    total: int


# ===== Learning Path V2 =====

class KnowledgePointRecordResponse(BaseModel):
    """知识点学习记录"""
    point_id: str
    point_name: str
    domain_name: str = ""
    subject_name: str = ""
    mastery_score: int = 0
    recent_accuracy: int = 0
    consecutive_errors: int = 0
    total_practiced: int = 0
    total_correct: int = 0
    total_questions: int = 0    # 该知识点在题库中的总题目数
    total_time_spent_seconds: int = 0
    study_count: int = 0
    last_study_at: Optional[datetime] = None
    last_practice_at: Optional[datetime] = None
    next_review_at: Optional[datetime] = None
    status: str = "not_started"
    video_url: Optional[str] = None       # 精讲视频链接
    review_material: Optional[str] = None  # 公共原文讲义（兼容旧字段名）
    review_source_url: Optional[str] = None
    review_source_mode: Optional[str] = None
    coding_problem_id: Optional[str] = None
    coding_problem_title: Optional[str] = None
    coding_problem_difficulty: Optional[str] = None
    model_config = {"from_attributes": True}


class PathNodeStatus(BaseModel):
    """路径中某个节点的状态"""
    point_id: str
    point_name: str
    domain_name: str = ""
    domain_sort_order: int = 0  # 领域排序序号
    sort_order: int = 0  # 知识点排序序号
    mastery_score: int = 0
    status: str = "not_started"  # not_started / learning / mastered / reviewing
    is_difficult: bool = False  # 困难点标记
    needs_review: bool = False  # 待复习标记


class DagNode(BaseModel):
    """DAG 节点（ReactFlow 格式）"""
    id: str
    point_id: str = ""
    label: str
    progress: str = "not_started"  # completed / in_progress / not_started
    mastery_score: int = 0
    is_weak: bool = False
    domain: str = ""
    subject: str = ""


class DagEdge(BaseModel):
    """DAG 边（ReactFlow 格式）"""
    id: str
    source: str
    target: str
    label: str = ""
    type: str = "PREREQUISITE"  # PREREQUISITE / RELATED_TO
    animated: bool = False


class DagData(BaseModel):
    """DAG 图数据"""
    nodes: List[DagNode] = []
    edges: List[DagEdge] = []
    metadata: dict = {}


class LearningPathMarkdownResponse(BaseModel):
    """学习路径响应：节点状态列表 + DAG 图数据"""
    nodes: List[PathNodeStatus]
    summary: dict = {}  # {total, mastered, learning, not_started, difficult}
    dag_data: DagData = Field(default_factory=DagData, description="ReactFlow DAG 图数据（含真实依赖边）")


class AgentRecommendation(BaseModel):
    """Agent 推荐卡片"""
    type: str  # review / practice / study_rest / unlock / breakthrough
    title: str
    description: str
    priority: str = "normal"  # high / normal / low
    related_point_id: Optional[str] = None
    related_point_name: Optional[str] = None
    action_label: str = "查看详情"
    action_url: str = ""


class AgentRecommendationListResponse(BaseModel):
    recommendations: List[AgentRecommendation]
    total: int


class PathHistoryItem(BaseModel):
    id: str
    agent_reason: Optional[str] = None
    snapshot_data: dict = {}
    created_at: datetime
    model_config = {"from_attributes": True}


class PathHistoryResponse(BaseModel):
    items: List[PathHistoryItem]
    total: int
